nexus/db-queries/src/db/datastore/reconfigurator_chicken_switches.rs

```diff
@@ -0,0 +1,351 @@
+        use nexus_db_schema::schema::reconfigurator_chicken_switches::dsl;
+
+        let latest = dsl::reconfigurator_chicken_switches
+            .filter(dsl::version.eq(SqlU32::new(version)))
+            .select(DbReconfiguratorChickenSwitches::as_select())
+            .get_result_async(&*conn)
+            .await
+            .optional()
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(latest.map(Into::into))
+    }
+
+    /// Insert the current version of the chicken switches in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently in
+    /// the `reconfigurator_chicken_switches` table. If there are no versions
+    /// currently in the table, then the current swtiches must be at version 1.
+    pub async fn reconfigurator_chicken_switches_insert_latest_version(
+        &self,
+        opctx: &OpContext,
+        switches: &ReconfiguratorChickenSwitches,
+    ) -> Result<(), Error> {
+        if switches.version < 1 {
+            return Err(Error::invalid_request(
+                "version must be greater than 0",
+            ));
+        }
+        opctx
+            .authorize(authz::Action::Modify, &authz::BLUEPRINT_CONFIG)
+            .await?;
+
+        let num_inserted = if switches.version == 1 {
+            self.reconfigurator_chicken_switches_insert_first(opctx, &switches)
+                .await?
+        } else {
+            self.reconfigurator_chicken_switches_insert_next_version(
+                opctx, &switches,
+            )
+            .await?
+        };
+
+        match num_inserted {
+            0 => Err(Error::invalid_request(format!(
+                "version {} is not the most recent",
+                switches.version
+            ))),
+            1 => Ok(()),
+            // This is impossible because we are explicitly inserting only one
+            // row with a unique primary key.
+            _ => unreachable!("query inserted more than one row"),
+        }
+    }
+
+    /// Insert the next version of the chicken switches in the database
+    ///
+    /// Only succeeds if the prior version is the latest version currently
+    /// in the `reconfigurator_chicken_switches` table.
+    ///
+    /// Panics if `switches.version <= 1`;
+    async fn reconfigurator_chicken_switches_insert_next_version(
+        &self,
+        opctx: &OpContext,
+        switches: &ReconfiguratorChickenSwitches,
+    ) -> Result<usize, Error> {
+        assert!(switches.version > 1);
+        let prev_version = switches.version - 1;
+
+        sql_query(
+            r"INSERT INTO reconfigurator_chicken_switches
+                (version, planner_enabled, time_modified)
+              SELECT $1, $2, $3
+              FROM reconfigurator_chicken_switches WHERE version = $4 AND version IN
+              (SELECT version FROM reconfigurator_chicken_switches
+               ORDER BY version DESC LIMIT 1)",
```

Comment:
I don't think we need a separate `insert_first` query. If we assume version 1 is the first, we can use COALESCE to claim there's a `0` if the table is empty, allowing the first insertion, while still doing the normal "only insert latest version + 1" if the table isn't empty:

```sql
INSERT INTO reconfigurator_chicken_switches
    (version, planner_enabled, time_modified)
SELECT $1, $2, $3
WHERE $3 - 1 IN (
    SELECT COALESCE(MAX(version), 0) FROM reconfigurator_chicken_switches
)
```

(Assuming `SELECT MAX(version)` uses the primary key index? Seems like it should be the same as `ORDER BY version DESC LIMIT 1` but I've been surprised before...)

Reply:
`COALESCE` looks great. I'll try it out. I'll actually need to also include the `WHERE version = $4` so I get compare and swap behavior on the previous version that's required. Hopefully that works with `COALESCE`.

Reply:
Oh, I see you already included that check with `$3 - 1`. Nice.

Reply:
Test passes. Pushed up in [6add8b6](https://github.com/oxidecomputer/omicron/pull/8449/commits/6add8b654900ee77ac3f34f2a1f5a72d12fffd57)